File: scalaShowandTell/src/main/scala/com/mp/showandtell/business/reader/CSVFileReader.scala
package com.mp.showandtell.business.reader

import java.io.{BufferedReader, File, FileReader, IOException}
import java.nio.charset.StandardCharsets
import java.nio.file.Path
import java.util

import com.mp.showandtell.business.ReaderInterface
import com.mp.showandtell.domain.{CreditInput, CreditMain}
import com.typesafe.scalalogging._
import org.springframework.stereotype.Component

import scala.io.Source

class CSVFileReader extends ReaderInterface {
  val logger = Logger(classOf[CSVFileReader])
  val DILIM_CSV: String = ","

  override def readFile(tempFile: Path): util.ArrayList[CreditInput] = {
    val output: util.ArrayList[CreditInput] = processCSVInputFile(tempFile)
    output
  }

  def processCSVInputFile(inputFilePath: Path): util.ArrayList[CreditInput] = {
    val file_Object = new File("abc.txt")
    println(file_Object.getAbsolutePath)
    val inputList = new util.ArrayList[CreditInput]()

    try {
      for (line <- Source.fromFile("Workbook2.csv").getLines) {
        println(line)
      }
    } catch {
      case exception: Exception =>
        println("Exception is Source parsing" + exception)
    }

    try {
      val br = new BufferedReader(new FileReader(inputFilePath.toString, StandardCharsets.ISO_8859_1))
      try { // skip the header of the csv
        var line1: String = null

        while (br.readLine() != null) {
          line1 = br.readLine();
          println("File : " + line1)
          inputList.add(mapToCredit(line1))
        }
      }
      catch {
        case e: IOException =>
          logger.error("File handling exception{}", e)
      } finally if (br != null) br.close()
    } catch {
      case e: IOException => logger.error("File handling exception{}", e)
    }
    inputList
  }

  def mapToCredit(splitV: String): CreditInput = {
    val p: Array[String] = splitV.split(DILIM_CSV)
    val creditObject = CreditMain.main(p)
    creditObject
  }
}

File: show-and-tell/src/main/java/com/mp/showandtell/business/reader/CSVFileReader.java
package com.mp.showandtell.business.reader;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mp.showandtell.business.FileRegister;
import com.mp.showandtell.business.ReaderInterface;
import com.mp.showandtell.domain.CreditInput;
import com.mp.showandtell.domain.CreditOutput;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.io.*;
import java.nio.charset.Charset;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

@Slf4j
@Component
public class CSVFileReader implements ReaderInterface {

    private static final String DILIM_CSV = ",";
    private static final ObjectMapper MAPPER = new ObjectMapper();

    static {
        FileRegister.register("csv", new CSVFileReader());
    }

    private Function<String, CreditInput> mapToItem = (line) -> {

        String[] p = line.split(DILIM_CSV);// a CSV has comma separated lines

        CreditInput input = new CreditInput();

        input.setName(p[0] + "," + p[1]);//<-- this is the first column in the csv file
        input.setAddress(p[2]);
        input.setPostCode(p[3]);
        if (p[4] != null && p[4].trim().length() > 0) {
            input.setPhoneNumber(p[4]);
        }
        if (p[5] != null && p[5].trim().length() > 0) {
            input.setCreditLimit(p[5]);
        }
        if (p[6] != null && p[6].trim().length() > 0) {
            input.setBirthDate(p[6]);
        }
        //more initialization goes here

        return input;
    };

    @Override
    public String readFile(Path tempFile) throws JsonProcessingException {

        return processCSVInputFile(tempFile.toString());
    }

    @Override
    public void writeFile() {
    }

    private String processCSVInputFile(String inputFilePath) throws JsonProcessingException {

        List<CreditInput> inputList = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(inputFilePath.toString(), Charset.forName("UTF-8")))){
           /* File inputF = new File(inputFilePath);
            InputStream inputFS = new FileInputStream(inputF);
            BufferedReader br = new BufferedReader(new InputStreamReader(inputFS));*/

            // skip the header of the csv
            inputList = br.lines().skip(1).map(mapToItem).collect(Collectors.toList());
//            br.close();
        } catch (IOException e) {
            log.error("File handling exception{}", e);
        }

        CreditOutput out = new CreditOutput();
        out.setOutput(inputList);
//        System.out.println(MAPPER.writeValueAsString(inputList));
        return MAPPER.writeValueAsString(out);
    }
}
